### PR TITLE
feat: [CDS-36921]: pipeline with env ref

### PR DESCRIPTION
**Pull request by AshwinBhatkal-Harness**
##### Summary:

1. Pipeline creation & update with env ref
2. Allow simultaneous deployments on Infra
3. Update of env and env group modal for creation from pipeline page

##### Jira Links:

https://harness.atlassian.net/browse/CDS-36921
https://harness.atlassian.net/browse/CDS-36922

##### Screenshots:

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

File: src/modules/70-pipeline/components/PipelineStudio/StepUtil.ts [hunk 1]
```diff
@@ -245,6 +245,8 @@ export const validateStage = ({
       }
     }
 
+    // TODO: errors
```
Comment (bhavya-sinha) [inline src/modules/70-pipeline/components/PipelineStudio/StepUtil.ts:248]: Please remember to remove this later

File: src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx [hunk 2]
```diff
@@ -93,8 +103,13 @@ export function InfrastructureModal({ hideModal, refetch, infrastructureToEdit,
             spec: {
               infrastructure: {
                 infrastructureDefinition: {
-                  type: defaultTo(type, deploymentTypeInfraTypeMap.Kubernetes),
-                  spec: defaultTo(spec, {})
+                  // type: defaultTo(type, deploymentTypeInfraTypeMap.Kubernetes),
+                  // spec: defaultTo(spec, {})
+                }
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:108]: Commented code?

Comment (bhavya-sinha) [inline src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:108]: Commented code?

File: src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx [hunk 3]
```diff
@@ -247,6 +271,20 @@ function BootstrapDeployInfraSpecifications({
       })
   }
 
+  const handleDeploymentTypeChange = useCallback(
+    (deploymentType: ServiceDeploymentType): void => {
+      if (deploymentType !== selectedDeploymentType) {
+        const stageData = produce(stage, draft => {
+          const serviceDefinition = get(draft, 'stage.spec.serviceConfig.serviceDefinition', {})
+          serviceDefinition.type = deploymentType
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:279]: Why are we updating serviceDefinition? This is not there in new specs

Comment (bhavya-sinha) [inline src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:279]: Why are we updating serviceDefinition? This is not there in new specs

File: src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx [hunk 4]
```diff
@@ -277,7 +320,12 @@ function BootstrapDeployInfraSpecifications({
                       }}
                     />
                   </Card>
-                  <DeployInfraSpecifications />
+                  <SelectDeploymentType
+                    selectedDeploymentType={selectedDeploymentType}
+                    isReadonly={false}
+                    handleDeploymentTypeChange={handleDeploymentTypeChange}
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:326]: I have made changes to add a new prop type here. https://github.com/harness/harness-core-ui/pull/9314

If my code gets merged first. Remember to update this

Comment (bhavya-sinha) [inline src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:326]: I have made changes to add a new prop type here. https://github.com/harness/harness-core-ui/pull/9314

If my code gets merged first. Remember to update this

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/AddEditEnvironmentModal.tsx [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+    },
+    {
+      label: getString('cd.preProduction'),
+      value: 'PreProduction'
+    }
+  ]
+  const formikRef = React.useRef<FormikProps<EnvironmentResponseDTO>>()
+  const id = data.identifier
+  const { data: environmentSchema } = useGetYamlSchema({
+    queryParams: {
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/AddEditEnvironmentModal.tsx:180]: Make sure this is updated schema. In service we have new schema.

Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/AddEditEnvironmentModal.tsx:180]: Make sure this is updated schema. In service we have new schema.

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    error,
+    refetch
+  } = useGetEnvironmentAccessList({
+    queryParams: { accountIdentifier: accountId, orgIdentifier, projectIdentifier },
+    lazy: true
+  })
+  const [environments, setEnvironments] = React.useState<SelectOption[]>([])
+
+  React.useEffect(() => {
+    refetch()
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx:93]: If we are refetching the api on mount, we neednot pass lazy: true.

**Review by bhavya-sinha — CHANGES_REQUESTED** [review]

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+            }}
+            disabled={inputSetData?.readonly}
+            className={css.inputWidth}
+            formik={formik}
+          />
+          {getMultiTypeFromValue(initialValues?.environmentRef) === MultiTypeInputType.FIXED && (
+            <Button
+              size={ButtonSize.SMALL}
+              variation={ButtonVariation.LINK}
+              disabled={inputSetData?.readonly || (isEditEnvironment(initialValues) ? !canEdit : !canCreate)}
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx:195]: We dont need usePermission Hook. Rbacbutton can be directly used here, which accepts permission props

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+      resourceType: ResourceType.ENVIRONMENT,
+      resourceIdentifier: environments[0]?.value as string
+    },
+    permissions: [PermissionIdentifier.EDIT_ENVIRONMENT],
+    options: {
+      skipCondition: ({ resourceIdentifier }) => !resourceIdentifier
+    }
+  })
+
+  const [canCreate] = usePermission({
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx:159]: This is not needed

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureWidget.tsx [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+                          size={ButtonSize.SMALL}
+                          variation={ButtonVariation.LINK}
+                          disabled={readonly || (isEditEnvironment(values) ? !canEdit : !canCreate)}
+                          onClick={() => {
+                            const isEdit = isEditEnvironment(values)
+                            if (isEdit) {
+                              //   if (values.environment?.identifier) {
+                              //     setState({
+                              //       isEdit,
+                              //       formik,
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureWidget.tsx:478]: Commented code?

File: src/modules/75-cd/components/PipelineStudio/DeployEnvSpecifications/DeployEnvSpecifications.tsx [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+            }}
+            allowableTypes={allowableTypes}
+            onUpdate={val => updateEnvStep(val)}
+            factory={factory}
+            stepViewType={StepViewType.Edit}
+          />
+        </Card>
+        {/* <SelectWithSubmenu
+          addClearBtn
+          onChange={itemSelect}
```
Comment (bhavya-sinha) [inline src/modules/75-cd/components/PipelineStudio/DeployEnvSpecifications/DeployEnvSpecifications.tsx:123]: Commented?

**Review by vkbansal — APPROVED** [review]

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureWidget.tsx (hunk 9 above)
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureWidget.tsx:478]: This is the environment group changes which is causing issues. Will put them back later

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureWidget.tsx [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+    [state]
+  )
+
+  const onClose = useCallback(() => {
+    setState({ isEdit: false, isEnvironment: false })
+    hideEnvironmentModal()
+  }, [hideEnvironmentModal])
+
+  const [canEdit] = usePermission({
+    resource: {
```
[the remark this answers is not in the document]
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureWidget.tsx:199]: I'll update it in the next PR

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+      setEnvironments(
+        environmentsResponse.data.map(env => ({
+          label: env.environment?.name || env.environment?.identifier || '',
+          value: env.environment?.identifier || ''
+        }))
+      )
+    }
+  }, [environmentsResponse, environmentsResponse?.data?.length])
+  const [canEdit] = usePermission({
+    resource: {
```
[the remark this answers is not in the document]
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx:149]: I'll update it in the next PR

File: src/modules/70-pipeline/components/PipelineStudio/StepUtil.ts (hunk 1 above)
Comment (AshwinBhatkal-Harness) [reply src/modules/70-pipeline/components/PipelineStudio/StepUtil.ts:248]: Yes

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx (hunk 8 above)
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx:159]: I'll update it in the next PR

File: src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx (hunk 6 above)
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/PipelineSteps/DeployInfrastructureStep/DeployInfrastructureInputStep.tsx:93]: I'll update it in the next PR

File: src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx (hunk 3 above)
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:279]: This is to load the deploymentType while creating infrastructure

File: src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx (hunk 2 above)
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/InfrastructureModal.tsx:108]: Will remove

**Review by bhavya-sinha — APPROVED** [review]

**Review by vkbansal — APPROVED** [review]

**Pull request merged**